plugin/options.go

```diff
@@ -0,0 +1,52 @@
+package plugin
+
+import (
+	"errors"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+// getOpts iterates the inbound Options and returns a struct
+func getOpts(opt ...Option) (*options, error) {
+	opts := getDefaultOptions()
+	for _, o := range opt {
+		if o == nil {
+			continue
+		}
+		iface := o()
+		switch to := iface.(type) {
+		case OptionFunc:
+			to(opts)
+		default:
+			return nil, errors.New("option passed in that does not belong to this package")
+		}
+	}
+	return opts, nil
+}
+
+type options struct {
+	withLogger hclog.Logger
+}
+
+// Option - a type that wraps an interface for compile-time safety but can
+// contain an option for this package or for wrappers implementing this
+// interface.
+type Option func() interface{}
+
+// OptionFunc - a type for funcs that operate on the shared Options struct. The
+// options below explicitly wrap this so that we can switch on it when parsing
+// opts for various wrappers.
+type OptionFunc func(*options)
+
+func getDefaultOptions() *options {
+	return &options{}
+}
+
+// WithLogger allows passing a logger to the plugin library for debugging
+func WithLogger(with hclog.Logger) Option {
+	return func() interface{} {
+		return OptionFunc(func(o *options) {
+			o.withLogger = with
+		})
+	}
```

Comment:
Nit:
```suggestion
func WithLogger(logger hclog.Logger) Option {
	return func() interface{} {
		return OptionFunc(func(o *options) {
			o.withLogger = logger
		})
	}
```

Reply:
This is my general pattern for options (although I've not been 100% consistent in the package -- I will update the base module to fix that later). Since it says "WithX", the parameter is the thing that it's with.

It's just a slightly more standardized way to do these. I don't feel strongly about it, but I also don't see a reason to change it.